core/shared/src/main/scala/zio/FiberRefs.scala

```diff
@@ -77,65 +77,62 @@ final class FiberRefs private (
    * preservation of maximum information from both child and parent refs.
    */
   def joinAs(fiberId: FiberId.Runtime)(that: FiberRefs): FiberRefs = {
+    @tailrec
+    def findAncestor(ref: FiberRef[Any])(
+      parentStack: List[(FiberId.Runtime, Any, Int)],
+      parentDepth: Int,
+      childStack: List[(FiberId.Runtime, Any, Int)],
+      childDepth: Int
+    ): Any =
+      (parentStack, childStack) match {
+        case (
+              (parentFiberId, parentValue, parentVersion) :: parentAncestors,
+              (childFiberId, childValue, childVersion) :: childAncestors
+            ) =>
+          if (parentFiberId == childFiberId)
+            if (childVersion > parentVersion) parentValue else childValue
+          else if (childDepth > parentDepth)
+            findAncestor(ref)(parentStack, parentDepth, childAncestors, childDepth - 1)
+          else if (childDepth < parentDepth)
+            findAncestor(ref)(parentAncestors, parentDepth - 1, childStack, childDepth)
+          else
+            findAncestor(ref)(parentAncestors, parentDepth - 1, childAncestors, childDepth - 1)
+        case _ =>
+          ref.initial
+      }
+
     val parentFiberRefs = self.fiberRefLocals
     val childFiberRefs  = that.fiberRefLocals
 
     val fiberRefLocals = childFiberRefs.foldLeft(parentFiberRefs) {
       case (parentFiberRefs, (fiberRef, (childStack, childDepth))) =>
-        val ref        = fiberRef.asInstanceOf[FiberRef[Any]]
-        val childValue = childStack.head._2
-
-        if (childStack.head._1 == fiberId) {
-          parentFiberRefs
-        } else {
+        if (childStack.head._1 == fiberId) parentFiberRefs
+        else {
+          val ref        = fiberRef.asInstanceOf[FiberRef[Any]]
+          val childValue = childStack.head._2
 
-          parentFiberRefs
-            .get(ref)
-            .fold {
+          parentFiberRefs.get(ref) match {
```

Comment:
Maybe `getOrNull` or equivalent, to avoid allocating an `Option`.

Reply:
Implemented